### ISSUE TITLE
Conflicting instantiation of the smart_manager Service model

### ISSUE THREAD

**Issue by FroggyFlox** [rockstor/rockstor-core#2723]
The `smart_manager.models.Service` model stores the services that are displayed and configurable (some) by the user on the *System* > *Services* page. We currently carry two different mechanisms to ensure that list is as it should be, and we do observe some interference between the two. Both of these mechanisms occur during Rockstor's initialization (`initrock.py`):

- SQL restore
- util function

# SQL restore
This mechanism occurs first, at:

https://github.com/rockstor/rockstor-core/blob/1269064c242b0261c61063944d1c8c7861a7cc66/src/rockstor/scripts/initrock.py#L87C34
```python
DB_SETUP["populate_smartdb"] = "psql smartdb -w -f {}/smartdb.sql.in".format(CONF_DIR)
```
"called" at:
https://github.com/rockstor/rockstor-core/blob/1269064c242b0261c61063944d1c8c7861a7cc66/src/rockstor/scripts/initrock.py#L554C18
```python
            run_command(["su", "-", "postgres", "-c", command])
```

This applies a dump of a past snapshot of that model table:
https://github.com/rockstor/rockstor-core/blob/1269064c242b0261c61063944d1c8c7861a7cc66/conf/smartdb.sql.in#L1343-L1365

This does set the table properly, but it makes it difficult to add new services or configs. For instance, the new Tailscale service (#2712) is not part of that db snapshot as it predates it.

> [!IMPORTANT]  
> This mechanism is only run at first Rockstor initialization (`STAMP` file not present). Rockstor RPM updates should not trigger it.

# Util function
This mechanism is defined at:
https://github.com/rockstor/rockstor-core/blob/1269064c242b0261c61063944d1c8c7861a7cc66/src/rockstor/scripts/prep_db.py#L23-L66

It is triggered at:
https://github.com/rockstor/rockstor-core/blob/1269064c242b0261c61063944d1c8c7861a7cc66/src/rockstor/scripts/initrock.py#L622-L624

> [!IMPORTANT]
> Contrary to the SQL restore mechanism, this one is called at every `initrock.py` run, regardless of the current db state.

The purpose of this function is to ensure the `Service` model includes all the entries it should by looping through a predefined list of services (list that should be used as source of truth?). If a service in the list is not present in the model, we create it.

During this creation, however, we currently have two bugs:

1. for services with a default config other than null set (shellinaboxd only at the moment), we override the config currently existing with the default. This means that every time `initrock.py` is run, we risk overwriting user changes. To test this, set a config for shellinabox that is different from default, reboot, and the user changes are gone, overwritten by default config.
2. If a service defined in the list (ground truth) is not in the model, we create it but miss the specification of the default config. This is currently not an issue as the only service with a default config is shellinabox, a service already created prior through the SQL restore mechanism. We should correct that bug, however.

# Proposal
We should harmonize this `Service` model instantiation.
The second method is more flexible and would allow us to add/remove services easily. Note, for instance, the presence of the now-removed netatalk service (AFP) in the SQL restore input that is then later removed by the second method (util function).

Maybe we can "simply" remove the `smartdb.sql.in` restore step? We should have all models already created in our smart_manager initial migration: `src/rockstor/smart_manager/migrations/0001_initial.py`

With regards to the two bugs listed above:

Bug 1: may be better fixed by seeing if we can move the Shellinabox service to have a config of null by default. In the short term, we shouldn't apply the default config if the service is already present in the model, especially considering the fact that the proper default config is already instantiated by method 1 (and thus assured to be there).

Bug 2: add the desired default config when creating the model instance should fix it.

At the exception of Bug 1, none of this represents a breaking behavior so not pressing. @phillxnet, how about we create a dedicated issue for Bug 1 with a temporary resolution while we fix the overall problem described in this issue (harmonizing our instantiation mechanisms for the `Service` model)?

**Comment by Hooverdan96** [comment]
Great findings. You think the SQL restore file is from the time before Django had its own migration functionality (and even before South was introduced)? Theoretically, like you said, this should not be necessary at all, right?

on a side note, would it make sense to squash migrations at every major release? Or would this make upgrades impossible/more complicated?

**Comment by phillxnet** [comment]
@FroggyFlox this is some nice archaeology. 
Re:

> ... how about we create a dedicated issue for Bug 1 with a temporary resolution while we fix the overall problem described in this issue (harmonizing our instantiation mechanisms for the `Service` model)?

Agreed. We then fix an observed bug and get that one out of the way, and normalised at least: ready to do the re-architecture.

@Hooverdan96 Re: 

> ... would it make sense to squash migrations at every major release? Or would this make upgrades impossible/more complicated?

Probably, but we did  that way back many years ago, and there was quite the long tail for upgrade from various places. Theoritically this is all good, and I've been tempted for ages to do this again, but we have to tread carefully as we try to support updating from as far back as we can. And we don't have many just yet. Some projects have hundreds, and we have gotten to over 50 in the past I think it was.

So I vote for not fixing that as it's not quite broken just yet :). Especially give we are changing so much - and I'm keen to be moving into the second half say of our testing phase so we can get the stable lot onto the new Python and the next Django (3.2) was my hope. Maye we then look again at a squash of all existing migrations, and try to simplify again our db instantiation/migration again; as we have had to add quite a few hacks to see what state it is in and try various things as we go. Working but messy. So yes, in time this would be good to have a new base. But last time it was painful to see the fall out. But granted things are getting better all the time migration wise, and yes we were using South for a bit: more stuff we can clean out once we finally squash what-ever our db is at some future Stable release say.

**Comment by phillxnet** [comment]
@FroggyFlox 
Re:

> Maybe we can "simply" remove the `smartdb.sql.in` restore step? We should have all models already created in our smart_manager initial migration: `src/rockstor/smart_manager/migrations/0001_initial.py`

Agreed. but wouldn't the initial migration only contain the services at that time. Follow-up migration files may add/remove there. I've not looked recently at this, but I remember when I added the Huey scheduler I just re-used ztaskd as it was way simpler and less invasive. And it was only a name :).

**Comment by FroggyFlox** [comment]
Thanks @Hooverdan96 for bringing these points out, and thanks @phillxnet for answering them! I was wondering the same things so that was very informative.

@phillxnet, I'll create an issue to address Bug 1 above on its own, then.

> Agreed. but wouldn't the initial migration only contain the services at that time. Follow-up migration files may add/remove there.

`src/rockstor/smart_manager/migrations/0001_initial.py` actually just creates the model without any entry and neither of the 2 following migrations seem to alter that for now. The population of the `Service` model seems to only be done through the 2 mechanisms that I described in the original post so it seems like an approach worth exploring.
Note that I have not looked at what's in the `smartdb.sql.in` dump for the other smart_manager models, though, so these may be a different story.

**Comment by phillxnet** [comment]
@FroggyFlox Re: my ongoing issues re initial setup in the linked:
"Update django-oauth-toolkit #2710" #2727
and:


> Maybe we can "simply" remove the `smartdb.sql.in` restore step?

I'm going to spin off from my current "Update django-oauth-toolkit" work, and from this issue regarding service and smartdb initialisation, the equivalent sql dump restore redundancy question; but for our default db of smartmanager. Not looked in detail yet but can then document within the issue as I look into it.

**Comment by phillxnet** [comment]
@FroggyFlox as you already know - but for the record/history on this issue - as of the referenced spin-off issue/pr (#2729/#2733) linked above we now no longer have any sql.in files (pg_dump generated) involved in our db instantiation mechanisms for either of our databases.

**Comment by FroggyFlox** [comment]
Closing this issue as it has been fixed as part of #2729/#2733 as indicated by @phillxnet.

In addition, 2 new issues were created to address the specific 2 bugs listed in the original post of this issue:
- #2724
- #2773
